### PR TITLE
[LowerSeqToSV] Use a dynamic index for array registers if possible

### PR DESCRIPTION
**Pull request by uenoku**
This PR tries lowering firrtl subaccess operation into SV variable
index assignment when possible. Because of the last connection sematnics,
it is not always possible to reconstruct subaccess operations. Specifically we can 
reconstruct subaccess only if there is no other write in the control flow path. Since bin flags are necessary for this transformation correctness, https://github.com/llvm/circt/pull/4689/commits/72c559ed67642ef5a466ea8fce5ace3c52bf2f07 propagates more bin flags thorough and op and icmp op canonicalizers.
```scala
circuit Example:
  module Example:
    input clock: Clock
    input en: UInt<1>
    input addr: UInt<2>
    input data: UInt<32>
    output out: UInt<32>[3]
    reg r : UInt<32>[3], clock

    when en:
      r[addr] <= data
    out <= r
```

```verilog
; $firtool -preserve-aggregate=1d-vec -preserve-public-types=false -disable-reg-randomization
module Mem(
  input              clock,
                     en,
  input  [1:0]       addr,
  input  [31:0]      data,
  output [2:0][31:0] out
);

  reg [2:0][31:0] r;
  always @(posedge clock) begin
    if (en)
      r[addr] <= data;
  end // always @(posedge)
  assign out = r;
endmodule
```

### CODE REVIEW

File: lib/Dialect/Comb/CombFolds.cpp [hunk 1]
```diff
@@ -3050,7 +3060,7 @@ LogicalResult ICmpOp::canonicalize(ICmpOp op, PatternRewriter &rewriter) {
                                            : APInt::getZero(width));
           replaceOpWithNewOpAndCopyName<ICmpOp>(rewriter, op, op.getPredicate(),
                                                 replicateOp.getInput(), cst,
-                                                false);
+                                                op.getTwoState());
```
Comment (darthscsi) [inline lib/Dialect/Comb/CombFolds.cpp:3063]: Can you split these two state preservations into a separate PR?

**Review by darthscsi — COMMENTED** [review]
can you try nested whens?

File: lib/Dialect/Comb/CombFolds.cpp (hunk 1 above)
Comment (uenoku) [reply lib/Dialect/Comb/CombFolds.cpp:3063]: Sure, https://github.com/llvm/circt/pull/4695 is the PR.

**Comment by uenoku** [comment]
> can you try nested whens?

Added a test generated from following firrtl:
```scala
circuit Mem:
  module Mem:
    input clock: Clock
    input en: UInt<1>[3]
    input addr: UInt<2>[4]
    input data: UInt<32>[4]
    output out: UInt<32>
    reg r : UInt<32>[3], clock

    when en[0]:
      when en[1]:
        r[addr[0]] <= data[0]
      else when en[2]:
        r[addr[1]] <= data[1]
      else:
        r[addr[2]] <= data[2]
    else:
      r[addr[3]] <= data[3]

    out <= r[addr[0]]
```

The output of `firtool  foo.fir -preserve-aggregate=1d-vec` is:
```verilog
  always @(posedge clock) begin
    if (en_0) begin
      if (en_1)
        r[addr_0] <= data_0;
      else if (en_2)
        r[addr_1] <= data_1;
      else
        r[addr_2] <= data_2;
    end
    else
      r[addr_3] <= data_3;
  end // always @(posedge)
```

**Comment by uenoku** [comment]
I'd like to merge the PR. It is only enabled with aggregate preservation flag right now. Happy to take a post commit review!
